consumergroup.go

```diff
@@ -0,0 +1,1048 @@
+// been assigned to a consumer.
+type PartitionAssignment struct {
+	// ID is the partition ID.
+	ID int
+
+	// Offset is the initial offset at which this assignment begins.  It will
+	// either be an absolute offset if one has previously been committed for
+	// the consumer group or a relative offset such as FirstOffset when this
+	// is the first time the partition have been assigned to a member of the
+	// group.
+	Offset int64
+}
+
+// genCtx adapts the done channel of the generation to a context.Context.  This
+// is used by Generation.Run so that we can pass a context to go routines
+// instead of passing around channels.
+type genCtx struct {
+	gen *Generation
+}
+
+func (c genCtx) Done() <-chan struct{} {
+	return c.gen.done
+}
+
+func (c genCtx) Err() error {
+	select {
+	case <-c.gen.done:
+		return ErrGenerationEnded
+	default:
+		return nil
+	}
+}
+
+func (c genCtx) Deadline() (time.Time, bool) {
+	return time.Time{}, false
+}
+
+func (c genCtx) Value(interface{}) interface{} {
+	return nil
+}
+
+// Generation represents a single consumer group generation.  The generation
+// carries the topic+partition assignments for the given.  It also provides
+// facilities for committing offsets and for running functions whose lifecycles
+// are bound to the generation.
+type Generation struct {
+	// ID is the generation ID as assigned by the consumer group coordinator.
+	ID int32
+
+	// GroupID is the name of the consumer group.
+	GroupID string
+
+	// MemberID is the ID assigned to this consumer by the consumer group
+	// coordinator.
+	MemberID string
+
+	// Assignments is the initial state of this Generation.  The partition
+	// assignments are grouped by topic.
+	Assignments map[string][]PartitionAssignment
+
+	conn coordinator
+
+	once sync.Once
+	done chan struct{}
+	wg   sync.WaitGroup
+
+	retentionMillis int64
+	log             func(func(*log.Logger))
+	logError        func(func(*log.Logger))
+}
+
+// close stops the generation and waits for all functions launched via Run to
+// terminate.
+func (g *Generation) close() {
+	g.once.Do(func() {
+		close(g.done)
+	})
+	g.wg.Wait()
+}
+
+// Run launches the provided function in a go routine and adds accounting such
+// that when the function exits, it stops the current generation (if not
+// already in the process of doing so).
+//
+// The provided function MUST support cancellation via the ctx argument and exit
+// in a timely manner once the ctx is complete.  When the context is closed, the
+// context's Error() function will return ErrGenerationEnded.
+//
+// When closing out a generation, the consumer group will wait for all functions
+// launched by Run to exit before the group can move on and join the next
+// generation.  If the function does not exit promptly, it will stop forward
+// progress for this consumer and potentially cause consumer group membership
+// churn.
+func (g *Generation) Run(fn func(ctx context.Context)) {
```

Comment:
```suggestion
func (g *Generation) Start(fn func(ctx context.Context)) {
```
I feel like `Run` is often used for blocking calls, which can be misleading to users of this package.